### PR TITLE
Add a snapshotting API and contextual transition example

### LINKED ISSUE #32: Add a snapshotting api
Something like this is fairly sufficient:

```objc

static UIView *FastSnapshotOfView(UIView *view) {
  return [view snapshotViewAfterScreenUpdates:NO];
}

static UIView *SlowSnapshotOfView(UIView *view) {
  UIGraphicsBeginImageContextWithOptions(view.bounds.size, NO, 0);
  [view.layer renderInContext:UIGraphicsGetCurrentContext()];
  UIImage *copied = UIGraphicsGetImageFromCurrentImageContext();
  UIView *copiedView = [[UIImageView alloc] initWithImage:copied];
  UIGraphicsEndImageContext();
  return copiedView;
}

  // Snapshotting

  NSMutableArray *hiddenViews = [NSMutableArray array];
  NSMutableArray *snapshotViews = [NSMutableArray array];
  UIView *(^snapshotOfView)(UIView *) = ^(UIView *view) {
    UIView *snapshotView = ((context.direction == MDMTransitionDirectionForward)
                            ? FastSnapshotOfView(view)
                            : SlowSnapshotOfView(view));

    snapshotView.frame = [context.containerView convertRect:view.bounds fromView:view];
    [context.containerView addSubview:snapshotView];
    [snapshotViews addObject:snapshotView];

    [hiddenViews addObject:view];
    view.hidden = YES;

    return snapshotView;
  };

  // Animations

  [CATransaction begin];
  [CATransaction setCompletionBlock:^{
    for (UIView *view in snapshotViews) {
      [view removeFromSuperview];
    }

    for (UIView *view in hiddenViews) {
      view.hidden = NO;
    }

    [context transitionDidEnd];
  }];
```

### PR DESCRIPTION
**Pull request by jverkoey**
Closes #32 and blocked by #36.

This API makes it easy for a transition engineer to create snapshots of views.

A new `deferToCompletion` API has been added to the transition context in order to facilitate cleanup logic in a transition. Any logic added via defer will be executed once the transition has ended.

```swift
let snapshotter = TransitionViewSnapshotter(containerView: context.containerView)
context.defer {
  snapshotter.removeAllSnapshots()
}

let snapshotView = snapshotter.snapshot(of: view,
                                        isAppearing: context.direction == .backward)
```

### CODE REVIEW

File: src/MDMTransitionContext.h [hunk 1]
```diff
@@ -83,4 +85,10 @@ NS_SWIFT_NAME(TransitionContext)
  The presentation view controller for this transition.
  */
 @property(nonatomic, strong, readonly, nullable) UIPresentationController *presentationController;
+
+/**
+ Defers execution of the provided work until the completion of the transition.
+ */
+- (void)deferToCompletion:(void (^ _Nonnull)())work;
```
Comment (ajsecord) [inline src/MDMTransitionContext.h]: Can I call this multiple times? What happens? Please document.

File: src/MDMTransitionViewSnapshotter.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ @param isAppearing If the view is appearing for the first time, a slower form of snapshotting may
+ be used. Otherwise, fast snapshotting may be used.
+ @return A new UIView instance that can be used as a visual replica of the provided view.
+ */
+- (nonnull UIView *)snapshotOfView:(nonnull UIView *)view isAppearing:(BOOL)isAppearing;
+
+/**
+ Removes all snapshots and unhides the snapshotted views.
+ */
+- (void)removeAllSnapshots;
```
Comment (ajsecord) [inline src/MDMTransitionViewSnapshotter.h]: Does this mean that this class owns the snapshots? Document ownership behavior.

**Review by ajsecord — CHANGES_REQUESTED** [review]

File: src/MDMTransitionContext.h (hunk 1 above)
Comment (jverkoey) [reply src/MDMTransitionContext.h]: Added some documentation.

File: src/MDMTransitionViewSnapshotter.h (hunk 2 above)
Comment (jverkoey) [reply src/MDMTransitionViewSnapshotter.h]: Added docs.

**Comment by jverkoey** [comment]
Gentle ping.

**Review by ajsecord — APPROVED** [review]

**Pull request merged**